Add Subscribe Id support (#305)
moq-wg/moq-transport: draft-ietf-moq-transport.md

specified track, as well as associated metadata required to deliver,
cache, and forward it. There are two subtypes of this message. When the
message type is 0x00, the optional Object Payload Length field is
present. When the message type is 0x02, the field is not present.

The format of the OBJECT message is as follows:

~~~
OBJECT Message {

>   Full Track Name Alias (i),

Reviewer: Individual Comment:

An alternate design is to put the Subscribe ID here, effectively merging it with Full Track Name Alias (I think this was Will's suggestion too).  Subscribe ID becomes an alias both for the name and the other subscribe params.  That simplifies a number of things, but the cost comes if we allow overlapping subscriptions to the same track and want to deduplicate objects among them.

Deduplicating overlapping subscriptions is a pain to the subscriber, who needs to track the ranges of their open subscriptions to know which callers want a given object, and for the server to deduplicate the objects before sending.

  Group Sequence (i),
  Object Sequence (i),
  Object Send Order (i),
  [Object Payload Length (i)],
  Object Payload (b),
}
~~~
{: #moq-transport-object-format title="MOQT OBJECT Message"}

> * Full Track Name Alias : The compressed full track name obtained as part of subscription and/or publish control message exchanges.

Reviewer 2: I'd prefer to keep this as Track ID for now, because it's so much shorter.  Or we could call it 'Track Alias' if that's slightly clearer?  Either way, let's discuss some other names on a PR and then we can do the rename in a separate PR from adding a Subscribe ID.

Reviewer 3: +1 to `Track Alias`

Author: +1 , i will update to

{: title="Relative Indexing"}

### SUBSCRIBE Format

The format of SUBSCRIBE is as follows:

~~~
SUBSCRIBE Message {
  Full Track Name Length (i),
  Full Track Name (...),

>   Subscribe ID (i),

Reviewer 3: IMO this should be the first field.

* Track Name: Identifies the track name as defined in ({{track-name}}).

> * SubscribeID: Session unique identifier for the subscription. `SubscribeID` MUST uniquely identify a subscription within a session and is used by subscribers and publishers to identify a given subscription. Subscribers generate the `SubscribeID` and MUST be

Reviewer: My recommendation is to be more prescriptive on subscribe ID assignment rather than just uniqueness - even/odd for client/server, monotonically increasing, cannot re-use IDs even after they have expired/closed.

> * SubscribeID: Session unique identifier for the subscription. `SubscribeID` MUST uniquely identify a subscription within a session and is used by subscribers and publishers to identify a given subscription. Subscribers generate the `SubscribeID` and MUST be

Reviewer 3: I don't think we need even/odd. Each publisher would have their own number space as they do currently.

But +1 to it being +1 for each SUBSCRIBE. It would be required for a cumulative MAX_SUBSCRIBES if go that direction.

nit: Can you add a space so its `Subscribe ID` instead of `SubscribeID`?

* Track Name: Identifies the track name as defined in ({{track-name}}).

* Subscribe ID: Session unique identifier for the subscription. `Subscribe ID`
is monotonically increasing variable length integer and it MUST uniquely
identify a subscription within a session. `Subscribe ID` is used by
subscribers and the publishers to identify a given subscription. Subscribers

> generate the `Subscribe ID` and it MUST be copied by the publisher when
> responding to the subscription requests.

Reviewer 2: ```suggestion
specify the `Subscribe ID` and it is included in the corresponding SUBSCRIBE_OK
or SUBSCRIBE_ERROR.
```

* Track Request Parameters: The parameters are defined in
{{version-specific-params}}

On successful subscription, the publisher SHOULD start delivering
objects from the group sequence and object sequence described above.

If a publisher cannot satisfy the requested start or end for the subscription it
MAY send a SUBSCRIBE_ERROR with code TBD. A publisher MUST NOT send objects
from outside the requested start and end.

> `SUBSCRIBE` with an existing `SubscribeID` updates the existing subscription and MUST NOT

Reviewer: Not sure you can enforce this MUST NOT - 'SHOULD NOT' or 'need not'?

Reviewer 3: I think we prohibit reusing `Subscribe IDs` until we flesh out resubscriptions in a separate PR.

Author: good point, i will remove this line and add it in the follow up PR i am working on

* Track Request Parameters: The parameters are defined in
{{version-specific-params}}

> 

Reviewer 2: ```suggestion
```

~~~
SUBSCRIBE_OK
{
  SubscribeID (i),
  Full Track Name Alias (i),
  Expires (i)
}
~~~
{: #moq-transport-subscribe-ok format title="MOQT SUBSCRIBE_OK Message"}

* SubscribeID: Subscription Identifer from the incoming subscription request for which this message is the response.

> * Full Track Name Alias: Session specific identifier that is used as an alias for the Full Track Name in the `Full Track Name Alias` field of the OBJECT ({{message-object}}) message headers of the requested track. `Full Track Name Alias` is generally shorter than Full Track Names and thus reduce the overhead in OBJECT messages.

Reviewer 3: Can you configure your editor to obey `.editorconfig` and word wrap at 80 characters? It's difficult to determine what changed.

  Track Namespace (b),
  Track Name (b),
  Track ID (i),

>   Subscribe ID (i),

Reviewer 3: IMO this should be the first field.

* Track ID: Session specific identifier that is used as an alias for the
Full Track Name in the Track ID field of the OBJECT ({{message-object}})
message headers of the requested track. Track IDs are shorter than
Full Track Names and thus reduce the overhead in OBJECT messages

> * Subscribe ID: Subscription Identifier specified in the corresponding SUBSCRIBE message.

Reviewer 2: ```suggestion
* Subscribe ID: Subscription Identifer as defined in {{message-subscribe-req}}.
```

* Track ID: Session specific identifier that is used as an alias for the
Full Track Name in the Track ID field of the OBJECT ({{message-object}})
message headers of the requested track. Track IDs are shorter than
Full Track Names and thus reduce the overhead in OBJECT messages

> * Subscribe ID: Subscription Identifer from the incoming subscription request
> for which this message is the response.

Reviewer 2: ```suggestion
* Subscribe ID: Subscription Identifier specified in the corresponding SUBSCRIBE message.
```

* Track ID: Session specific identifier that is used as an alias for the
Full Track Name in the Track ID field of the OBJECT ({{message-object}})

> message headers of the requested track. Track IDs are generally shorter than

Reviewer 2: ```suggestion
message headers of the requested track. Track IDs are shorter than
```

~~~
SUBSCRIBE_FIN Message {
  SubscribeID (i),
  Final Group (i),
  Final Object (i),
}
~~~
{: #moq-transport-subscribe-fin-format title="MOQT SUBSCRIBE_FIN Message"}

> * SubscribeID: Identifies the subscription.

Reviewer 3: This definition is useless. Can you at least include a link back to the `SUBSCRIBE` section?
